File: src/routes/trades.ts
import { Router, Request, Response } from "express";
import { authMiddleware } from "../middleware/auth";

export const tradeRouter = Router();

// ---- 我的交换历史 ----
tradeRouter.get("/", authMiddleware, async (req: Request, res: Response) => {
  const { prisma } = await import("../index");
  const trades = await prisma.tradeOffer.findMany({
    where: {
      status: "completed",
      OR: [{ proposerId: req.user!.userId }, { receiverId: req.user!.userId }],
    },
    include: {
      listing: {
        include: {
          offeredItem: { include: { definition: true } },
        },
      },
      proposer: { select: { id: true, username: true, avatarUrl: true } },
      receiver: { select: { id: true, username: true, avatarUrl: true } },
      offerItems: {
        include: { inventoryItem: { include: { definition: true } } },
      },
    },
    orderBy: { createdAt: "desc" },
  });
  res.json({ trades });
});

// ---- 交换详情 ----
tradeRouter.get("/:id", authMiddleware, async (req: Request, res: Response) => {
  const { prisma } = await import("../index");
  const trade = await prisma.tradeOffer.findUnique({
    where: { id: BigInt(req.params.id) },
    include: {
      listing: {
        include: {
          offeredItem: { include: { definition: true } },
          seller: { select: { id: true, username: true } },
        },
      },
      proposer: { select: { id: true, username: true, avatarUrl: true } },
      receiver: { select: { id: true, username: true, avatarUrl: true } },
      offerItems: {
        include: { inventoryItem: { include: { definition: true } } },
      },
      eventLogs: { orderBy: { createdAt: "asc" } },
    },
  });
  if (!trade) { res.status(404).json({ error: "交换记录不存在" }); return; }
  res.json({ trade });
});

File: app/api/trades.py
"""交换历史路由"""
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session, joinedload
from app.database import get_db
from app.models import TradeOffer, OfferStatus, OfferItem, InventoryItem
from app.schemas import TradeOfferOut
from app.deps import get_current_user_id
from app.errors import NotFound

router = APIRouter(prefix="/api/trades", tags=["交换历史"])


@router.get("", response_model=list[TradeOfferOut])
def list_trades(
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """我的交换历史"""
    trades = db.query(TradeOffer).options(
        joinedload(TradeOffer.listing),
        joinedload(TradeOffer.proposer),
        joinedload(TradeOffer.receiver),
        joinedload(TradeOffer.offer_items).joinedload(OfferItem.inventory_item).joinedload(InventoryItem.definition),
    ).filter(
        TradeOffer.status == OfferStatus.completed,
        (TradeOffer.proposer_id == user_id) | (TradeOffer.receiver_id == user_id),
    ).order_by(TradeOffer.created_at.desc()).all()

    return [TradeOfferOut.from_orm(t) for t in trades]


@router.get("/{offer_id}", response_model=TradeOfferOut)
def get_trade(
    offer_id: int,
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """交换详情"""
    trade = db.query(TradeOffer).options(
        joinedload(TradeOffer.listing),
        joinedload(TradeOffer.proposer),
        joinedload(TradeOffer.receiver),
        joinedload(TradeOffer.offer_items).joinedload(OfferItem.inventory_item).joinedload(InventoryItem.definition),
    ).filter(TradeOffer.id == offer_id).first()

    if not trade:
        raise NotFound("交换记录不存在")

    return TradeOfferOut.from_orm(trade)
